test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
-	rsV1beta2, err := clientSet.AppsV1beta2().ReplicaSets(ns.Name).Get(replicaset.Name, metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("failed to get apps/v1beta2 replicaset %s: %v", replicaset.Name, err)
+	markAllPodsRunningAndReady(t, c, pods, ns.Name, *rs.Spec.Replicas)
+	waitAllPodsReady(t, c, rs, ns.Name)
+	if rs.Status.AvailableReplicas != 0 {
+		t.Fatalf("Unexpected .Status.AvailableReplicas: Expected 0, saw %d", rs.Status.AvailableReplicas)
 	}
-	newSelectorLabels = map[string]string{"changed_name_apps_v1beta2": "changed_test_apps_v1beta2"}
-	rsV1beta2.Spec.Selector.MatchLabels = newSelectorLabels
-	rsV1beta2.Spec.Template.Labels = newSelectorLabels
-	_, err = clientSet.AppsV1beta2().ReplicaSets(ns.Name).Update(rsV1beta2)
-	if err == nil {
-		t.Fatalf("failed to provide validation error when changing immutable selector when updating apps/v1beta2 replicaset %s", rsV1beta2.Name)
+
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// Make .spec.minReadySeconds = 0 so that we can test the replicas are available
+		newRS.Spec.MinReadySeconds = 0
+		_, err = rsClient.Update(newRS)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update .Spec.MinReadySeconds of rs %s: %v", rs.Name, rs.Status.AvailableReplicas)
 	}
-	expectedErrType := "Invalid value"
-	expectedErrDetail := "field is immutable"
-	if !strings.Contains(err.Error(), expectedErrType) || !strings.Contains(err.Error(), expectedErrDetail) {
-		t.Errorf("error message does not match, expected type: %s, expected detail: %s, got: %s", expectedErrType, expectedErrDetail, err.Error())
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if newRS.Status.AvailableReplicas != *newRS.Spec.Replicas {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatalf("Unexpected .Status.AvailableReplicas: Expected 3, saw %d", rs.Status.AvailableReplicas)
+	}
+}
+
+func testScalingUsingScaleEndpoint(t *testing.T, c clientset.Interface, rs *v1beta1.ReplicaSet, ns string, x int32) {
+	kind := "ReplicaSet"
+	scaleClient := c.ExtensionsV1beta1().Scales(ns)
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		scale, err := scaleClient.Get(kind, rs.Name)
+		if err != nil {
+			return err
+		}
+		scale.Spec.Replicas = x
+		_, err = scaleClient.Update(kind, scale)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to set .Spec.Replicas of scale subresource for rs %s: %v", rs.Name, err)
+	}
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newRS, err := c.Extensions().ReplicaSets(ns).Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if newRS.Status.Replicas != x {
```

Comment:
A better alternative is to get the scale subresource and then `newRS.Status.Replicas = scale.Spec.Replicas`.

Reply:
No, that could hide a bug in which scale.Spec.Replicas is not what we expect. It's better to expect the value `x` (which should be renamed `replicas` by the way) because that comes authoritatively from the caller.

Reply:
It's not really necessary to check that Status.Replicas converges to the desired value. We test elsewhere that RS creates Pods and updates Status correctly. What we want to test here is just that setting Scale.Spec.Replicas causes RS.Spec.Replicas to be updated.

Instead, I recommend checking more aspects of the mapping between the RS and Scale objects. For example, let the caller input an expected value of Scale.Spec.Replicas *before* the update, so you can verify information flow from RS->Scale object. Alternatively, you could get this expected value by doing a fresh GET on the RS at the beginning of this function.

Reply:
Ya, checking whether `Status.Replicas` converges to desired value is unnecessary.
I will focus on verifying information does flow from RS to its Scale object. Good summary.